docs/config-reference.md
# Configuration reference

The default bundle configuration:

```yaml
# Default configuration for extension with alias: "huh_pdf_creator"
huh_pdf_creator:

    # Set to true to use this bundle functionality in the contao article syndication.
    enable_contao_article_pdf_syndication: false

    # Set the paths allowed for file opening (e.g. image loading). Paths are be relativ to the project dir. Currently only used for dompdf (chroot).
    allowed_paths:

        # Defaults:
        - web
        - public
        - files
        - assets

    # PDF creator configurations
    configurations:

        # Prototype: The title of the configuration. Should be a unique alias/name containing just 'a-z0-9-_' like 'news_export','default','brand_a_themed'.
        title:

            # The pdf creator type (pdf library).
            type:                 ~ # One of "dompdf"; "mpdf"; "tcpdf"

            # A nice name for displaying in the backend.
            name:                 ~ # Example: 'A4 brochure with corporate logo'

            # Set a file name for the generated pdf files. You can use the %title% placeholder to use the title of the content to export in the file name.
            filename:             '%%title%%.pdf'

            # The path to the folder where the generated files should be stored. Only used if output_mode is AbstractPdfCreator::OUTPUT_MODE_FILE. Path must be relative to the project path.
            file_path:            ~ # Example: files/export/pdf

            # Set page orientation.
            orientation:          portrait # One of "portrait"; "landscape"

            # Set how to output the pdf.
            output_mode:          inline # One of "download"; "file"; "inline"; "string"

            # Set a page format. This could be a standardized format like A3, A4, A5 or Legal, otherwise you can specify the format in millimeter (width x height, seperated by comma, for example 180,210).
            format:               A4
            margins:

                # Relative path from project to font file.
                top:                  ~

                # Name of the font family
                left:                 ~

                # Font style
                bottom:               ~

                # Font weight
                right:                ~
                unit:                 mm # One of "mm"
            fonts:

                # Prototype
                -

                    # Relative path from project to font file.
                    path:                 ~

                    # Name of the font family
                    family:               ~

                    # Font style
                    style:                ~

                    # Font weight
                    weight:               ~

            # Set a pdf template (also known as master template), which will be the base template for the generated pdf files. Must be a path relative to the contao web root.
            base_template:        ~ # Example: files/media/news/news_base_template.pdf

            # Set a default url that will override the url that is determined by the request. This can be usefull on development servers with custom url mapping.
            base_url:             null # Example: 'https://stage.example.org:8001/examplepath/'

            # Set credentials for basic http authentication.
            credentials:          null # Example: 'user:password'
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.6] - 2023-06-29
- Added: allowed_paths configuration option for dompdf chroot setting [(#6)]
- Added: default chroot paths for dompdf [(#6)]
- Changed: some internal refactoring [(#6)]
- Fixed: missing dependencies

## [0.4.5] - 2023-06-02
- Fixed: exception in syndication typ

## [0.4.4] - 2023-05-09
- Added: allow passing listener through pdf generator context
- Fixed: warning

## [0.4.3] - 2023-05-05
- Changed: allow psr/log 2 and 3

## [0.4.2] - 2022-11-21
- Fixed: warnings under php 8

## [0.4.1] - 2022-07-27
- Fixed: page margin right config option name
- Fixed: font support from yaml configuration

## [0.4.0] - 2022-07-26
- Added: filePath option to configuration for pdf file output type ([#4])
- Added: margins and font options to yaml configuration ([#4])
- Added: PdfGeneratorContext override config property ([#4])
- Changed: minimum contao version is now 4.9 ([#4])
- Changed: minimum pdf creator version is now 0.4 ([#4])
- Changed: removed utils bundle dependency ([#4])
- Changed: removed utils bundle dependency ([#4])
- Fixed: symfony 5 related issues ([#4])
- Fixed: missing php dependency ([#4])

## [0.3.2] - 2022-03-10
- Changed: updated configuration documentation
- Fixed: used print encore entries instead of own encore entries in syndication type

## [0.3.1] - 2021-06-30
- fixed PdfGenerator::getConfiguration() not always returned null if config not found
- fixed PdfCreatorSyndicationType::export() not throws exception if export configuration not found

## [0.3.0] - 2021-05-25
- Add option to configure pdf creator from yaml ([#3])
- add base_url and credentials option (only yaml) ([#3])
- add PdfCreatorModel to the BeforeCreateLibraryInstanceEvent and BeforeOutputPdfCallbackEvent ([#3])
- fixed Dompdf log in prod mode ([#3])
- adjustments due changes in pdf creator library 0.3.2

## [0.2.2] - 2021-04-20
- increased pdf creator library dependency to 0.3
- added support for Dompdf (translations and custom debug implementation)
- added support for feature support check to hide functionality not provided by a pdf creator class

## [0.2.1] - 2021-04-06
- fixed missing syndication type translation

## [0.2.0] - 2021-03-09
- syndication type now use twig
- syndication type now supports encore bundle

## [0.1.1] - 2021-03-05
- updated the syndication type

## [0.1.0] - 2021-02-26
Initial release

[#3]: https://github.com/heimrichhannot/contao-pdf-creator-bundle/pull/3
[#4]: https://github.com/heimrichhannot/contao-pdf-creator-bundle/pull/4
[#6]: https://github.com/heimrichhannot/contao-pdf-creator-bundle/pull/6


src/DependencyInjection/Configuration.php
<?php

/*
 * Copyright (c) 2023 Heimrich & Hannot GmbH
 *
 * @license LGPL-3.0-or-later
 */

namespace Heimrichhannot\PdfCreatorBundle\DependencyInjection;

use HeimrichHannot\PdfCreator\AbstractPdfCreator;
use HeimrichHannot\PdfCreator\PdfCreatorFactory;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder('huh_pdf_creator');
        $treeBuilder->getRootNode()
            ->children()
                ->booleanNode('enable_contao_article_pdf_syndication')
                    ->defaultFalse()
                    ->info('Set to true to use this bundle functionality in the contao article syndication.')
                ->end()
                ->arrayNode('allowed_paths')
                    ->info('Set the paths allowed for file opening (e.g. image loading). Paths are be relativ to the project dir. Currently only used for dompdf (chroot).')
                    ->scalarPrototype()->end()
                    ->defaultValue(['web', 'public', 'files', 'assets'])
                ->end()
                ->append($this->pdfCreatorConfigurations())
            ->end()
        ;

        return $treeBuilder;
    }

    private function pdfCreatorConfigurations()
    {
        $treeBuilder = new TreeBuilder('configurations');

        $node = $treeBuilder->getRootNode()
            ->info('PDF creator configurations')
            ->useAttributeAsKey('title')
            ->arrayPrototype()
                ->info("The title of the configuration. Should be a unique alias/name containing just 'a-z0-9-_' like 'news_export','default','brand_a_themed'.")
                ->children()
                    ->enumNode('type')
                        ->values(PdfCreatorFactory::getTypes())
                        ->info('The pdf creator type (pdf library).')
                    ->end()
                    ->scalarNode('name')
                        ->info('A nice name for displaying in the backend.')
                        ->example('A4 brochure with corporate logo')
                    ->end()
                    ->scalarNode('filename')
                        ->defaultValue('%%title%%.pdf')
                        ->info('Set a file name for the generated pdf files. You can use the %title% placeholder to use the title of the content to export in the file name.')
                    ->end()
                    ->scalarNode('file_path')
                        ->info(
                            'The path to the folder where the generated files should be stored. '.
                            'Only used if output_mode is AbstractPdfCreator::OUTPUT_MODE_FILE. '.
                            'Path must be relative to the project path.'
                        )
                        ->example('files/export/pdf')
                    ->end()
                    ->enumNode('orientation')
                    ->values([AbstractPdfCreator::ORIENTATION_PORTRAIT, AbstractPdfCreator::ORIENTATION_LANDSCAPE])
                    ->defaultValue(AbstractPdfCreator::ORIENTATION_PORTRAIT)
                    ->info('Set page orientation.')
                    ->end()
                    ->enumNode('output_mode')
                    ->values(AbstractPdfCreator::OUTPUT_MODES)
                    ->defaultValue(AbstractPdfCreator::OUTPUT_MODE_INLINE)
                    ->info('Set how to output the pdf.')
                    ->end()
                    ->scalarNode('format')
                    ->defaultValue('A4')
                    ->info('Set a page format. This could be a standardized format like A3, A4, A5 or Legal, otherwise you can specify the format in millimeter (width x height, seperated by comma, for example 180,210).')
                    ->end()
                    ->arrayNode('margins')
                        ->children()
                            ->scalarNode('top')->info('Relative path from project to font file.')->end()
                            ->scalarNode('left')->info('Name of the font family')->end()
                            ->scalarNode('bottom')->info('Font style')->end()
                            ->scalarNode('right')->info('Font weight')->end()
                            ->enumNode('unit')->defaultValue('mm')->values(['mm'])->end()
                        ->end()
                    ->end()
                    ->arrayNode('fonts')
                        ->arrayPrototype()
                            ->children()
                                ->scalarNode('path')->info('Relative path from project to font file.')->end()
                                ->scalarNode('family')->info('Name of the font family')->end()
                                ->scalarNode('style')->info('Font style')->end()
                                ->scalarNode('weight')->info('Font weight')->end()
                            ->end()
                        ->end()
                    ->end()
                    ->scalarNode('base_template')
                        ->info('Set a pdf template (also known as master template), which will be the base template for the generated pdf files. Must be a path relative to the contao web root.')
                        ->example('files/media/news/news_base_template.pdf')
                    ->end()
                    ->scalarNode('base_url')
                        ->defaultNull()
                        ->info('Set a default url that will override the url that is determined by the request. This can be usefull on development servers with custom url mapping.')
                        ->example('https://stage.example.org:8001/examplepath/')
                    ->end()
                    ->scalarNode('credentials')
                        ->defaultNull()
                        ->info('Set credentials for basic http authentication.')
                        ->example('user:password')
                    ->end()
                ->end()
            ->end()
        ;

        return $node;
    }
}
